### DISCUSSION TITLE
[jsvm] Help creating a signed s3 URL?

### DISCUSSION THREAD

**Discussion by benallfree** [pocketbase/pocketbase#5983] [Q&A] [answered]
I am trying to serve s3 files using a 302 to signed s3 URLs similar to #4260, but using JSVM.

Signing requires an HMAC chain and I'm having trouble creating it.

Here is the example in nodejs that works:

```js
// WORKS
import crypto from 'crypto'

const foo = crypto.createHmac('sha256', 'secret').update('first-value').digest()
console.log(`foo`, foo.toString('hex')) // fb59a8ad6a4bccb9ad2fd6719ba6ea49ead1ff246988f099cf7e74bbf5ef6d79
const bar = crypto.createHmac('sha256', foo).update('second-value').digest()
console.log(`bar`, bar.toString('hex')) // 21ed0471f66291002d0fc1871450c193cd93ee676d6dac97d07b4bdd55667e74
```

I also confirmed the same works in Go:

```go
// WORKS
	foo := hmac.New(sha256.New, []byte("secret"))
	foo.Write([]byte("first-value"))
	fooDigest := foo.Sum(nil)
	fmt.Printf("foo: %x\n", fooDigest) // fb59a8ad6a4bccb9ad2fd6719ba6ea49ead1ff246988f099cf7e74bbf5ef6d79

	bar := hmac.New(sha256.New, fooDigest)
	bar.Write([]byte("second-value"))
	barDigest := bar.Sum(nil)
	fmt.Printf("bar: %x\n", barDigest) // 21ed0471f66291002d0fc1871450c193cd93ee676d6dac97d07b4bdd55667e74
```

And here is the same in JSVM, but it doesn't work:

```js
// NOT WORKING
  const foo = $security.hs256(`first-value`, 'secret')
  console.log(`foo`, foo) // fb59a8ad6a4bccb9ad2fd6719ba6ea49ead1ff246988f099cf7e74bbf5ef6d79
  const bar = $security.hs256('second-value', foo) 
  console.log(`bar`, bar) // e41097eb5606277834eafe4a1fbbb39c4282a322592d6f33953134b956629d4c
```

I have also tried converting `foo` in various ways such as:

```js
  const hexToBytes = (hex) =>
    hex
      .match(/.{2}/g)
      .map((v) => String.fromCharCode(parseInt(v, 16)))
      .join('')
```

But that also does not work.

How do I make an HMAC chain in JSVM?

**Comment by chnfyi** [comment]
It's because your hexToBytes function returns a string while nodejs hmac `digest()` returns a Buffer that "_is a subclass of JavaScript's Uint8Array_" ([nodejs](https://nodejs.org/api/crypto.html#hmacdigestencoding)) when no encoding is provided.

So you can try this instead:
```javascript
const hexToBytes = (hex) =>
  new Uint8Array(hex.match(/.{2}/g).map((v) => parseInt(v, 16)))
```

**Reply by benallfree** [reply]
Nope :( 

```js
  {
    const hexToBytes = (hex) =>
      Buffer.from(hex.match(/.{2}/g).map((v) => parseInt(v, 16)))
    const bar = $security.hs256('second-value', hexToBytes(foo))
    console.log(`bar`, bar) // 1d45800e7bff6dfa84e402b31b21a222734b3fc3e41115ae143ccc82902aabc8
  }
  {
    const hexToBytes = (hex) =>
      new Uint8Array(hex.match(/.{2}/g).map((v) => parseInt(v, 16)))
    const bar = $security.hs256('second-value', hexToBytes(foo))
    console.log(`bar`, bar) // a870d7dc7fe9cd2d24f9c5339a73011f44e27f06d2f99b2daeef1519a12fdb53
  }
  {
    const hexToBytes = (hex) => hex.match(/.{2}/g).map((v) => parseInt(v, 16))
    const bar = $security.hs256('second-value', hexToBytes(foo))
    console.log(`bar`, bar) // a870d7dc7fe9cd2d24f9c5339a73011f44e27f06d2f99b2daeef1519a12fdb53
  }
  {
    const hexToBytes = (hex) =>
      hex
        .match(/.{2}/g)
        .map((v) => String.fromCharCode(parseInt(v, 16)))
        .join('')
    const bar = $security.hs256('second-value', hexToBytes(foo))
    console.log(`bar`, bar) // 57e31fae5e952fe5180fba02bfa1a1b742d9fa88e42d856c8690bdd7a65e3a04
  }
```

**Comment by benallfree** [comment]
Interesting note... if I modify `HS256` to recognize hex secrets, it works. 

```go
func HS256(text string, secret string) string {
	// Try to decode hex string if possible
	if decoded, err := hex.DecodeString(secret); err == nil {
		secret = string(decoded)
	}
	h := hmac.New(sha256.New, []byte(secret))
	h.Write([]byte(text))
	return fmt.Sprintf("%x", h.Sum(nil))
}
```

But there seems to be no way to pass in a binary string from goja.

**Comment by benallfree** [comment]
https://github.com/dop251/goja/issues/631

**Reply by ganigeorgiev** [reply, marked as answer]
I don't think this is goja issue since the equivalent Node.js code behaves identical:

```js
// Node.js
const foo = crypto.createHmac('sha256', 'secret').update('first-value').digest()
const bar = crypto.createHmac('sha256', foo.toString()).update('second-value').digest()
console.log(bar) // 1d45800e7bff6dfa84e402b31b21a222734b3fc3e41115ae143ccc82902aabc8

// PocketBase
const foo = $security.hs256('first-value', 'secret')
const bar = $security.hs256('second-value', new Buffer(foo, "hex").toString())
console.log(bar) // 1d45800e7bff6dfa84e402b31b21a222734b3fc3e41115ae143ccc82902aabc8
```

The problem is most likely caused by the fact that the PocketBase binding operates only with string secret VS Node.js example using Buffer (_unicode characters can be represented in multiple different ways which could lead to bytes sequence differences during the argument type conversion_).

A workaround could be to perform the conversion on the Go side by calling the `toString(bytes)` which internally will do just `string(bytes)`. Or in other words, try with:

```js
const foo = $security.hs256('first-value', 'secret')
const bar = $security.hs256('second-value', toString(new Buffer(foo, "hex")))
console.log(bar) // 21ed0471f66291002d0fc1871450c193cd93ee676d6dac97d07b4bdd55667e74
```

I'll consider with the next release to update the bindings to accept `[]byte` argument instead of string since technically it is more correct and will minimize the conversion caveats between JS and Go.

**Reply by benallfree** [reply]
Are you referring to [toString](https://github.com/dop251/goja/blob/master/builtin_string.go#L27)? Can you please provide more insight on why `toString(new Buffer())` works, but `new Buffer().toString()` does not? These types of JSVM quirks are something I'm still struggling to fully understand.

**Reply by ganigeorgiev** [reply]
No, `toString` is a [PocketBase global helper](https://github.com/pocketbase/pocketbase/blob/master/plugins/jsvm/binds.go#L301-L343) (_in hindsight maybe it wasn't a good idea to name it such_).

It was intended to replace the old `readerToString` as a more generic approach when dealing with mutltibytes characters conversions as in https://github.com/pocketbase/pocketbase/discussions/4488#discussioncomment-10458781, json marshalization, etc.

I haven't explored how the `Buffer` goja module works but again as shown in the first example its `.toString()` method behaves identical to the Node.js version so I don't think it is goja issue.

**Reply by benallfree** [reply]
1. Is there a <=v0.22 equivalent of `toString(new Buffer(foo, 'hex'))`? It says `Buffer` is not defined
2. How are `toString(new Buffer())` and `Buffer.toString()` different?

**Reply by benallfree** [reply]
Note: another workaround that is compatible with both v22 and v23 is https://github.com/Caligatio/jsSHA:

```js
  const JSSHA = require(`${__hooks}/jssha256`)

  const hs256 = (data, key, keyFormat = 'HEX') => {
    const sha = new JSSHA('SHA-256', 'TEXT', {
      hmacKey: { value: key, format: keyFormat },
    })
    sha.update(data)
    return sha.getHash('HEX')
  }
```

**Reply by ganigeorgiev** [reply]
1. The `Buffer` module is enabled only in PocketBase v0.23+. An equivalent I guess would be anything that decodes the hex string into `[]byte`, for example you can try with one of the earlier suggestions:
    ```js
    const bar = $security.hs256('second-value', toString(
        new Uint8Array(foo.match(/.{2}/g).map((v) => parseInt(v, 16)))
    ))
    ```
2. The goja types specific `toString()` method is not the same as the generic Go helper and its implementation would vary from one type to another while I imagine trying to be as close as possible to its ES5 or Node.js implementation. From a brief look at the [Buffer source code](https://github.com/dop251/goja_nodejs/blob/master/buffer/buffer.go#L391) it seems that it does a lot more than just calling the Go Stringer interface method or doing `string(bytes)`.
